log-levels.md

```diff
@@ -0,0 +1,159 @@
+# Log Levels 
+
+This guide serves as guidelines for library authors with regards to what [SwiftLog](https://github.com/apple/swift-log) log levels are apropriate for use in libraries, and in what siuations to use what level.
+
+## Guidelines for Libraries
+
+SwiftLog defines the following 7 log levels via the [`Logger.Level` enum](https://apple.github.io/swift-log/docs/current/Logging/Structs/Logger/Level.html), ordered from least to most severe: 
+
+* `trace`
+* `debug`
+* `info`
+* `notice`
+* `warning`
+* `error`
+* `critical`
+
+Out of those, only levels _less severe than_ info (exclusive) are generally okay to be used by libraries.
+In the following section we'll explore how to use them in practice.
+
+### Recommended log levels
+
+It is always fine for a library to log at `trace` and `debug` levels, and these two should be the primary levels any library is logging at.
+
+`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug". It is expected to take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
+
+This is in contrast with `debug` which some users _may_ choose to run with enabled on their production systems.
+
+> Debug level logging should be not "too" noisy. Developers should assume some production deployments may need to (or want to) run with debug level logging enabled. 
+>
+> Debug level logging should not completely undermine the performance of a production system.
+
+As such, `debug` logging should not be overly noisy. It should provide a high value understanding of what is going on in the library for end users without diving deep into its internals. This is what `trace` is intended for.
+
+Use `warning` level sparingly. Whenever possible, try to rather emit errors to end users that are descriptive enough so they can inspect, log them and figure out the issue. Potentially they may then enable debug logging to find out more about the issue.
```

Comment:
I think we could be somewhat more precise around "emit errors". I think what you mean is instead of logging an error at `warning`, throw the error/fail the future with an appropriate error message that the user can then log (or not).

Reply:
yes, rephrased it to be more precise